File: framework-ts/module/devian-domain-game/features/ability/AbilityUnitBase.ts
import { UNIT_STAT_TYPE } from '../../Generated/Game.g';
import { AbilityBase } from './AbilityBase';

export abstract class AbilityUnitBase extends AbilityBase {
    private mCurHP: number = 0;

    abstract get unitId(): string;
    get unitLevel(): number { return this.getInt(UNIT_STAT_TYPE.UNIT_LEVEL); }
    get curHP(): number { return this.mCurHP; }

    protected initUnitState(unitLevel: number, maxHp: number): void {
        this.setStat(UNIT_STAT_TYPE.UNIT_LEVEL, unitLevel);
        this.setStat(UNIT_STAT_TYPE.UNIT_HP, maxHp);
        this.mCurHP = this.maxHP;
    }

    protected copyUnitStateFrom(source: AbilityUnitBase): void {
        this.mCurHP = source.mCurHP;
    }
}

File: framework-cs/apps/UnityExample/Assets/Samples/Devian Foundation/0.1.0/GamePackage/Runtime/Ability/AbilityUnitBase.cs
using Devian.Domain.Game;

namespace Devian
{
    public abstract class AbilityUnitBase : AbilityBase
    {
        public abstract string UnitId { get; }
        int mCurHP = 0;

        public int UnitLevel => this[UNIT_STAT_TYPE.UNIT_LEVEL];
        public int CurHP => mCurHP;

        protected void InitUnitState(int unitLevel, int maxHp)
        {
            SetStat(UNIT_STAT_TYPE.UNIT_LEVEL, unitLevel);
            SetStat(UNIT_STAT_TYPE.UNIT_HP, maxHp);
            mCurHP = MaxHP;
        }

        protected void CopyUnitStateFrom(AbilityUnitBase source)
        {
            mCurHP = source.mCurHP;
        }

        internal void ResetCurrentHpToMax()
        {
            mCurHP = MaxHP;
        }
    }
}
